### PR TITLE
Handle missing virtual-router or listener gracefully to be backward compatible

### PR DESCRIPTION
**Pull request by kiranmeduri**
*Issue #, if available:*
#63 

*Description of changes:*
Non backward behavior was introduced to virtual-service CRD that adds a new required field virtual-router as dictated by App Mesh API. This broke existing CRD and examples elsewhere lingering on the web. It is possible to update all the examples but it is not favorable to break existing setups when controller is upgraded without any notice or warning. 

Any users using latest controller (v0.1.2+) with old CRDs will face this problem and will be unable to create virtual-services.

This fix allows creating virtual-routers without specifying listeners. Controller now looks at the routes associated with virtual-service to determine the suitable listener from the virtual-nodes that are registered as targets for those routes.

*Testing*
- Added unit test that covers all the edge cases
- Manual e2e test using local app with following spec
```
apiVersion: appmesh.k8s.aws/v1beta1
kind: VirtualService
metadata:
  name: color.${APP_NAMESPACE}.svc.cluster.local
  namespace: ${APP_NAMESPACE}
spec:
  meshName: ${MESH_NAME}
  routes:
    - name: color-route-blue
      priority: 10
      http:
        match:
          prefix: /
          headers:
            - name: color_header
              match:
                exact: blue
        action:
          weightedTargets:
            - virtualNodeName: blue
              weight: 1
```
Results in an update resource spec 
```
...
  virtualRouter:
    listeners:
    - portMapping:
        port: 8080
        protocol: http
    name: color.howto-k8s-http-headers.svc.cluster.local
status:
  conditions:
  - lastTransitionTime: "2019-09-17T23:34:47Z"
    status: "True"
    type: VirtualRouterActive
  - lastTransitionTime: "2019-09-17T23:34:48Z"
    status: "True"
    type: RoutesActive
  - lastTransitionTime: "2019-09-17T23:34:50Z"
    status: "True"
    type: VirtualServiceActive
```
Corresponding log lines
```
I0917 23:34:47.121709       1 virtualservice.go:354] Setting router listener to {{8080 http} 0xc000ed2440} for virtual-service  color.howto-k8s-http-headers.svc.cluster.local in namespace howto-k8s-http-headers
I0917 23:34:47.359274       1 virtualnode.go:107] Updated virtual node green-howto-k8s-http-headers
E0917 23:34:47.722645       1 virtualservice.go:340] Error getting virtual-node with name blue-howto-k8s-http-headers in namespace howto-k8s-http-headers, virtualnodes.appmesh.k8s.aws "blue-howto-k8s-http-headers" not found
I0917 23:34:47.731124       1 appmesh.go:643] Using 0 vrouter listeners to build 0 input listeners
E0917 23:34:47.731399       1 controller.go:443] error syncing 'howto-k8s-http-headers/color.howto-k8s-http-headers.svc.cluster.local': error creating virtual router: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, CreateVirtualRouterInput.Spec.Listeners.
I0917 23:34:47.747134       1 appmesh.go:643] Using 1 vrouter listeners to build 1 input listeners
I0917 23:34:47.770272       1 virtualservice.go:114] Created virtual router color-howto-k8s-http-headers-svc-cluster-local
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controller/virtualservice.go [hunk 1]
```diff
@@ -284,13 +284,76 @@ func getVServiceCondition(conditionType appmeshv1beta1.VirtualServiceConditionTy
...
+		return nil
 	}
-	return &appmeshv1beta1.VirtualRouter{
-		Name: vservice.Name,
+
+	var candidateVnodeName string
+	route := vservice.Spec.Routes[0]
+	if route.Http != nil {
+		if len(route.Http.Action.WeightedTargets) > 0 {
+			candidateVnodeName = route.Http.Action.WeightedTargets[0].VirtualNodeName
```
[the remark this answers is not in the document]
Comment (kiranmeduri) [reply pkg/controller/virtualservice.go]: yeah given that it is edge case we can probably error out if virtual-nodes are using different port-mapping. That said it can be deep check that can be inefficient. Only sane thing is to slowly migrate to defining listener on virtual-router.

[the remark this answers is not in the document]
Comment (bcelenza) [reply pkg/controller/virtualservice.go]: Yeah, my main concern is if someone changes the order of the Virtual Nodes in the route, and those nodes have a different listener port, they'll have broken the downstream clients, potentially without knowing.

[the remark this answers is not in the document]
Comment (kiranmeduri) [reply pkg/controller/virtualservice.go]: alternately #74 may be enough to do.

[the remark this answers is not in the document]
Comment (bcelenza) [reply pkg/controller/virtualservice.go]: Sounds good. Thanks!

**Comment by jerrygb** [comment]
@kiranmeduri @bcelenza Can you please confirm the current examples on AWS docs are still broken assuming there is no new release after v0.1.2?

The examples still point to the following image.

``
602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon/app-mesh-controller:v0.1.2
```

**Comment by jerrygb** [comment]
Nvm, I see that you have the updates coming in soon @ https://github.com/aws/aws-app-mesh-controller-for-k8s/pull/78
